### PR TITLE
Improve handling of archives when installing assets

### LINKED ISSUE #75712: Editor treats custom repository host zips differently.
I noticed that some plug-ins that have their repository host set as custom seem to have their zips treated differently.  For example, some plug-ins I have downloaded, the editor seems to ignore the base folders and just dump the contents into the project's root.  However, ones that have, say, Github listed as the repository host with a commit hash to get the zips, those unzip correctly.

In both cases, the editor seems to see the folders.  It just doesn't install correctly for some reason.  Not sure if this is intended, a bug, or what.

### LINKED ISSUE #81311: Addon is empty after downloading from asset lib
### Godot version

v4.1.1.stable.official [bd6af8e0e]

### System information

Godot v4.1.1.stable - Windows 10.0.22621 - Vulkan (Forward+) - dedicated NVIDIA GeForce GTX 1080 Ti (NVIDIA; 31.0.15.3667) - AMD Ryzen 9 7900X 12-Core Processor (24 Threads)

### Issue description

This [addon](https://godotengine.org/asset-library/asset/1989) is shown as empty when downloaded via the editor. If downloading from the website everything works fine and there is content in the zip file.
image

The asset uses `Custom` as repository host, linking with URL to `https://github.com/kenyoni-software/godot-addons/releases/download/hide_private_properties-1.0.1/hide_private_properties-1_0_1.zip`

### Steps to reproduce

Go to the Asset Lib and download [Hide Private Properties](https://godotengine.org/asset-library/asset/1989).

### Minimal reproduction project

N/A

### PR DESCRIPTION
**Pull request by YuriSizov**
Fixes https://github.com/godotengine/godot/issues/81311. Closes https://github.com/godotengine/godot/issues/75712.

So https://github.com/godotengine/godot/issues/75712 has been a problem for quite some time now. As discussed, there are two key points:
- Installing ZIPs manually is treated differently than installing assets from the AssetLib (despite handling the same ZIPs internally);
- All ZIPs are expected to have a root directory in them, because that's how GitHub, etc pack their ZIPs.

And https://github.com/godotengine/godot/issues/81311 is a new finding, as far as I know. It seems some ZIP archives can skip the directory structure information and just contain record of files without intermediate folders. We fail at handling those gracefully.

-----

So I set to fix these problems, and along the way I tweaks smaller details here and there. Here's a list of changes:

* Implicitly create folders when parsing ZIP files to make sure we have the entire structure available regardless of the source ZIP.
* Add a checkbox that allows to skip the root directory when extracting files, _if such directory exists_.
* Fix formatting of the "Files failed to extract" error message that was missing a line break.
* Fix some errors not being displayed at all (also remove the unnecessary dialog and use toasts instead).
* Improve the status bar to better highlight the title of the asset and the status on the conflicting files (also use `TTRN` for that last message).
* Rename the window to explain what it's for.
* Various refactoring for code clarity and slightly better organization (this is clearly a very old file, and it leaves a lot to be desired, so I kept the changes to rather minimal).

godot windows editor dev x86_64_2023-09-05_23-54-53
godot windows editor dev x86_64_2023-09-05_23-54-28

godot windows editor dev x86_64_2023-09-06_00-05-32

Most of it should be self explanatory. Regarding the checkbox, it tries to be smart. First of all, it's disabled if there is no single root folder in the asset archive. It only becomes available if your asset has a folder structure similar to what GitHub creates. This may backfire when your ZIP file contains, for example, only the `addons` folder — but if that becomes an issue we should be able to easily add an exception.

I've decided to preserve the current dual logic, where assets installed through the AssetLib have their root folder skipped, while archives installed manually don't behave this way. However, this is no longer a hardcoded behavior. Instead, the checkbox is either checked by default or not, but can still be changed by the user.

Here's everything in action (note that the image loading error is related to the broken asset and is not a part of this PR):

https://github.com/godotengine/godot/assets/11782833/5e5add57-130a-4a06-b82e-09b4a4e2c4a8

### CODE REVIEW

**Comment by Trey2k** [comment]
Would it be possible to modify what info the editor retries from the asset library to verify if the repo host is custom or not? I think that could potentially be a cleaner approach as then we would know for certain how it should be handled in either case.

**Comment by YuriSizov** [comment]
A custom host doesn't mean the asset doesn't follow this structure. And a custom host can still be used to point to a GitHub archive.

Giving users control is, IMO, the best option.

File: editor/editor_asset_installer.cpp [hunk 1]
```diff
@@ -82,229 +81,259 @@ void EditorAssetInstaller::open(const String &p_path, int p_depth) {
...
-		files_sorted.insert(name);
+		String source_name = String::utf8(fname);
 
-		ret = unzGoToNextFile(pkg);
-	}
+		// Create intermediate directories if they aren't reported by unzip.
+		// We are only interested in subfolders, so skip the root slash.
+		int separator = source_name.find("/", 1);
+		while (separator != -1) {
+			String dir_name = source_name.substr(0, separator) + "/";
```
Comment (akien-mga) [inline editor/editor_asset_installer.cpp:90]: Nit: Could likely use `split("/")` instead of manual `find` and `substr`, but it's not important.

File: editor/editor_asset_installer.cpp [hunk 2]
```diff
@@ -82,229 +81,259 @@ void EditorAssetInstaller::open(const String &p_path, int p_depth) {
...
+}
+
+void EditorAssetInstaller::_check_has_toplevel() {
+	// Check if the file structure has a distinct top-level directory. This is typical
+	// for archives generated by GitHub, etc, but not for manually created ZIPs.
+
+	toplevel_prefix = "";
+	skip_toplevel_check->set_pressed(false);
+	skip_toplevel_check->set_disabled(true);
+	skip_toplevel_check->set_tooltip_text(TTR("This asset doesn't have one root directory, so it can't be ignored."));
```
Comment (akien-mga) [inline editor/editor_asset_installer.cpp:236]:
```suggestion
	skip_toplevel_check->set_tooltip_text(TTR("This asset doesn't have a root directory, so it can't be ignored."));
```

File: editor/editor_asset_installer.cpp [hunk 3]
```diff
@@ -338,26 +367,102 @@ String EditorAssetInstaller::get_asset_name() const {
...
 	VBoxContainer *vb = memnew(VBoxContainer);
 	add_child(vb);
 
-	asset_contents = memnew(Label);
-	vb->add_child(asset_contents);
+	HBoxContainer *asset_status = memnew(HBoxContainer);
+	vb->add_child(asset_status);
+
+	Label *asset_label = memnew(Label);
+	asset_label->set_text(TTR("Contents of asset:"));
```
Comment (akien-mga) [inline editor/editor_asset_installer.cpp:440]:
```suggestion
	asset_label->set_text(TTR("Asset contents:"));
```
maybe?

**Review by akien-mga — APPROVED** [review]
Looks fine to me from a very cursory code review, and the PR rationale sounds good.

Could likely use a more in depth review/testing by @KoBeWi or @Calinou.

File: editor/editor_asset_installer.cpp (hunk 3 above)
Comment (akien-mga) [reply editor/editor_asset_installer.cpp:440]: Nevermind, I saw the screenshot and it's better like this.

File: editor/editor_asset_installer.cpp [hunk 4]
```diff
@@ -338,26 +367,102 @@ String EditorAssetInstaller::get_asset_name() const {
...
 	tree = memnew(Tree);
 	tree->set_v_size_flags(Control::SIZE_EXPAND_FILL);
-	tree->connect("item_edited", callable_mp(this, &EditorAssetInstaller::_item_edited));
+	tree->connect("item_edited", callable_mp(this, &EditorAssetInstaller::_item_checked));
 	tree->connect("check_propagated_to_item", callable_mp(this, &EditorAssetInstaller::_check_propagated_to_item));
 	vb->add_child(tree);
 
-	error = memnew(AcceptDialog);
-	add_child(error);
+	set_title(TTR("Select asset files to install"));
```
Comment (akien-mga) [inline editor/editor_asset_installer.cpp:465]: As much as I personally dislike it, our convention for titles is to use Title Case.
```suggestion
	set_title(TTR("Select Asset Files to Install"));
```

File: editor/editor_asset_installer.cpp (hunk 1 above)
Comment (KoBeWi) [reply editor/editor_asset_installer.cpp:90]: Or better, `get_slice()` and `get_slice_count()`.

**Comment by fire** [comment]
@lyuma this was a big problem for our extensions and offers alternative for our problem of requiring the root to be project root and not directory addons root

File: editor/editor_asset_installer.cpp (hunk 1 above)
Comment (KoBeWi) [reply editor/editor_asset_installer.cpp:90]: It's actually better that way in this case. If anything, it could be replaced with `get_slice()`.

File: editor/editor_asset_installer.cpp (hunk 2 above)
Comment (YuriSizov) [reply editor/editor_asset_installer.cpp:236]: I tried to underline that it's looking for the one single root folder, but if it's still unclear (or clear even with your suggestion), then I can make the change.

**Comment by YuriSizov** [comment]
> this was a big problem for our extensions and offers alternative for our problem of requiring the root to be project root and not directory addons root

I'm not sure this PR would solve your issue. But after this little refactor I'm quite confident I can implement the "export to this folder" feature, so archives can be exported to places other than the root.

**Comment by IceflowRE** [comment]
Tested regarding https://github.com/godotengine/godot/issues/81311, and it works now.

File: editor/editor_asset_installer.h [hunk 5]
```diff
@@ -33,25 +33,49 @@
 
 #include "scene/gui/dialogs.h"
 #include "scene/gui/tree.h"
+
+class CheckBox;
+class Label;
```
Comment (KoBeWi) [inline editor/editor_asset_installer.h:38]: Why was this added if Label was already used without problems?

File: editor/editor_asset_installer.cpp [hunk 6]
```diff
@@ -82,229 +81,259 @@ void EditorAssetInstaller::open(const String &p_path, int p_depth) {
...
+void EditorAssetInstaller::_check_has_toplevel() {
+	// Check if the file structure has a distinct top-level directory. This is typical
+	// for archives generated by GitHub, etc, but not for manually created ZIPs.
+
+	toplevel_prefix = "";
+	skip_toplevel_check->set_pressed(false);
+	skip_toplevel_check->set_disabled(true);
+	skip_toplevel_check->set_tooltip_text(TTR("This asset doesn't have one root directory, so it can't be ignored."));
+
+	if (asset_files.size() == 0) {
```
Comment (KoBeWi) [inline editor/editor_asset_installer.cpp:238]:
```suggestion
	if (asset_files.is_empty()) {
```

File: editor/editor_asset_installer.cpp [hunk 7]
```diff
@@ -82,229 +81,259 @@ void EditorAssetInstaller::open(const String &p_path, int p_depth) {
...
+			if (!toplevel_prefix.is_empty() && skip_toplevel) {
+				continue;
 			}
-			path = path.substr(pp + 1, path.length());
-			depth--;
 		}
 
-		if (skip || path.is_empty()) {
-			continue;
+		String path = E; // We're going to mutate it.
```
Comment (KoBeWi) [inline editor/editor_asset_installer.cpp:139]: You could just remove `const` from E.

File: editor/editor_asset_installer.cpp [hunk 8]
```diff
@@ -82,229 +81,259 @@ void EditorAssetInstaller::open(const String &p_path, int p_depth) {
...
+		int separator = path.rfind("/");
+		if (separator == -1) {
 			parent_item = root;
 		} else {
-			String ppath = path.substr(0, pp);
-			ERR_CONTINUE(!dir_map.has(ppath));
-			parent_item = dir_map[ppath];
+			String parent_path = path.substr(0, separator);
+			ERR_CONTINUE(!directory_item_map.has(parent_path));
+			parent_item = directory_item_map[parent_path];
```
Comment (KoBeWi) [inline editor/editor_asset_installer.cpp:159]:
```suggestion
			HashMap<String, TreeItem *>::Iterator I = directory_item_map.find(parent_path);
			ERR_CONTINUE(!I);
			parent_item = I->value();
```

File: editor/editor_asset_installer.cpp [hunk 9]
```diff
@@ -82,229 +81,259 @@ void EditorAssetInstaller::open(const String &p_path, int p_depth) {
...
+			unzCloseCurrentFile(pkg);
 
-				Ref<FileAccess> f = FileAccess::open(path, FileAccess::WRITE);
-				if (f.is_valid()) {
-					f->store_buffer(uncomp_data.ptr(), uncomp_data.size());
-				} else {
-					failed_files.push_back(path);
-				}
+			// Ensure that the target folder exists.
+			Ref<DirAccess> da = DirAccess::create(DirAccess::ACCESS_RESOURCES);
```
Comment (KoBeWi) [inline editor/editor_asset_installer.cpp:325]: I think you could create this before the loop.

File: editor/editor_asset_installer.cpp (hunk 7 above)
Comment (YuriSizov) [reply editor/editor_asset_installer.cpp:139]: We don't want to mutate the contents of the container.

File: editor/editor_asset_installer.h (hunk 5 above)
Comment (YuriSizov) [reply editor/editor_asset_installer.h:38]: Just a precaution because these are used without a problem implicitly. I'm not sure dialogs and tree should guarantee these to be included (likely could benefit from forward declarations as well).

File: editor/editor_asset_installer.cpp (hunk 7 above)
Comment (KoBeWi) [reply editor/editor_asset_installer.cpp:139]: Ah, then it's fine.
I don't think it would create copy, though not sure.

File: editor/editor_asset_installer.cpp (hunk 1 above)
Comment (YuriSizov) [reply editor/editor_asset_installer.cpp:90]: Not sure that split would be an improvement, we would still need to iterate over the collection. `get_slice`/`get_slice_count` also seem to be less efficient for this, I think. But I agree with the final suggestion, it makes sense.

**Comment by YuriSizov** [comment]
All concerns should be addressed, I think.

File: editor/editor_asset_installer.h (hunk 5 above)
Comment (KoBeWi) [reply editor/editor_asset_installer.h:38]: There are many more classes included explicitly.

**Review by KoBeWi — APPROVED** [review]

**Comment by akien-mga** [comment]
Thanks!
